openstudiocore/src/model/WaterHeaterHeatPump.cpp

```diff
@@ -0,0 +1,834 @@
+    return OS_WaterHeater_HeatPumpFields::AirOutletNode;
+  }
+
+  ModelObject WaterHeaterHeatPump_Impl::clone(Model model) const
+  {
+    auto newWaterHeater = ZoneHVACComponent_Impl::clone(model).cast<WaterHeaterHeatPump>();
+
+    {
+      auto mo = tank().clone(model).cast<HVACComponent>();
+      newWaterHeater.setTank(mo);
+    }
+
+    {
+      auto mo = dXCoil().clone(model).cast<ModelObject>();
+      newWaterHeater.setDXCoil(mo);
+    }
+
+    {
+      auto mo = fan().clone(model).cast<HVACComponent>();
+      newWaterHeater.setFan(mo);
+    }
+
+    return newWaterHeater;
+  }
+
+  std::vector<ModelObject> WaterHeaterHeatPump_Impl::children() const
+  {
+    std::vector<ModelObject> result;
+
+    result.push_back(tank());
+    result.push_back(dXCoil());
+    result.push_back(fan());
+
+    return result;
+  }
+
+  bool WaterHeaterHeatPump_Impl::addToThermalZone(ThermalZone & thermalZone)
+  {
+    bool result = false;
+    auto thisObject = getObject<WaterHeaterHeatPump>();
+
+    if( (result = ZoneHVACComponent_Impl::addToThermalZone(thermalZone)) ) {
+      thermalZone.setHeatingPriority(thisObject,1);
+      thermalZone.setCoolingPriority(thisObject,1);
+      setCompressorLocation("Zone");
+      setInletAirConfiguration("ZoneAirOnly");
+      auto t_tank = tank();
+      if( auto waterHeaterMixed = t_tank.optionalCast<WaterHeaterMixed>() ) {
+        waterHeaterMixed->setAmbientTemperatureIndicator("ThermalZone");
+        waterHeaterMixed->setAmbientTemperatureThermalZone(thermalZone);
+      }
+    }
+
+    return result;
+  }
+
+} // detail
+
+WaterHeaterHeatPump::WaterHeaterHeatPump(const Model& model,
+  const ModelObject & dxCoil,
+  const HVACComponent & tank,
+  const HVACComponent & fan,
+  Schedule & compressorSetpointTemperatureSchedule,
+  Schedule & inletAirMixerSchedule)
+  : ZoneHVACComponent(WaterHeaterHeatPump::iddObjectType(),model)
+{
+  OS_ASSERT(getImpl<detail::WaterHeaterHeatPump_Impl>());
+
+  setDXCoil(dxCoil);
+  setTank(tank);
+  setFan(fan);
+  setCompressorSetpointTemperatureSchedule(compressorSetpointTemperatureSchedule);
+  setInletAirMixerSchedule(inletAirMixerSchedule);
+
+  setDeadBandTemperatureDifference(5.0);
+  autosizeCondenserWaterFlowRate();
+  autosizeEvaporatorAirFlowRate();
+  setInletAirConfiguration("Schedule");
+  setMinimumInletAirTemperatureforCompressorOperation(10.0);
+  setCompressorLocation("Schedule");
+  setFanPlacement("DrawThrough");
+  setOnCycleParasiticElectricLoad(0.0);
+  setOffCycleParasiticElectricLoad(0.0);
+  setParasiticHeatRejectionLocation("Outdoors");
+  setControlSensorLocationInStratifiedTank("Heater1");
+}
+
+WaterHeaterHeatPump::WaterHeaterHeatPump(const Model& model)
+  : ZoneHVACComponent(WaterHeaterHeatPump::iddObjectType(),model)
+{
+  OS_ASSERT(getImpl<detail::WaterHeaterHeatPump_Impl>());
+
+  CoilWaterHeatingAirToWaterHeatPump coil(model);
+  setDXCoil(coil);
+
+  WaterHeaterMixed waterHeater(model);
```

Comment:
I kept the ::setTank ::tank methods generic with HVACComponent in anticipation.  @mbadams5 it might be good to start with the stratified tank when you get going.